Switch from pygmsh to the gmsh Python API

jorgensd opened the pull request:
As pygmsh has started using the gmsh python API from version 7.0.0, there is no point adding an extra dependency and compatibility layer between gmsh and dolfinx.

Meshio has also been deprecated from the docker images. It is not required to run demos or tests.
A direct interface from msh->dolfinx has been added, using the GMSH API to read in the mesh on one processor and distribute the data using dolfin-x.

Also worthwhile discussing whether the functions introduced in `demo_gmsh.py` should be part of `dolfinx.io`.

Demo rewritten using only GMSH python API. Related to #1143

garth-wells commented on docker/Dockerfile:
```diff
@@ -145,12 +145,12 @@ RUN export DEBIAN_FRONTEND=noninteractive && \
     export HDF5_MPI="ON" && \
     export CC=mpicc && \
-    pip3 install --no-cache-dir --no-binary=h5py h5py meshio pygmsh
+    pip3 install --no-cache-dir --no-binary=h5py h5py meshio
```
Do we still need `meshio`?

jorgensd replied:
It is not a strict dependency any longer. However, to interface with msh-files, we would require it. Maybe we should create a separate PR with a msh->dolfin demo.

garth-wells replied:
Not sure I see the point. I would remove it if it isn't a test dependency. The few dependencies, the fewer things can can break a build.

jorgensd replied:
We need to have an interface to `msh` files at some point, as it is quite strict to require everyone to go through the `gmsh` interface.
In many cases, people might have msh files from other software.
The main problem is that meshio `msh`->`xdmf` is not directly compatible with dolfinx, see: https://fenicsproject.discourse.group/t/transitioning-from-mesh-xml-to-mesh-xdmf-from-dolfin-convert-to-meshio/412/79?u=dokken

garth-wells commented on python/demo/gmsh/demo_gmsh.py:
```diff
@@ -5,33 +5,101 @@
 # =====================================
-# Mesh generation using Gmsh and pygmsh
+# Mesh generation using the GMSH python API
 # =====================================
```
Fix the rst `=` markers.

garth-wells commented on python/demo/gmsh/demo_gmsh.py:
```diff
@@ -5,33 +5,101 @@
+import numpy as np
+
+
+def gather_gmsh_physical_entities():
```
This function loops really complicated. Can it be simplified? 

At the lest, removing some line breaks and adding come comments would improve readability.

jorgensd replied:
We need this function to gather physical tags into meshes. For instance when using `BooleanFragements` with gmsh,  you get multiple volume entities that has to be loaded.
This splitting is very common with entities of lower dimensions, such as facets.
I can add in some comments to explain what is going on.

jorgensd replied:
Function has been slightly simplified and is now well documented. 
Should we make it part of the dolfinx.io?

garth-wells commented on python/demo/gmsh/demo_gmsh.py:
```diff
@@ -5,33 +5,101 @@
+    srt = np.argsort(idx)
+    assert np.all(idx[srt] == np.arange(len(idx)))
+    x = points[srt]
+    return x
```
`return points[srt]`

garth-wells commented on python/demo/gmsh/demo_gmsh.py:
```diff
@@ -5,33 +5,101 @@
+    idx, points, _ = gmsh.model.mesh.getNodes()
+    points = points.reshape(-1, 3)
+    idx -= 1
+    srt = np.argsort(idx)
```
`srt` is cryptoc.

garth-wells commented on python/demo/gmsh/demo_gmsh.py:
```diff
@@ -5,33 +5,101 @@
+    return meshes
+
+
+def sort_gmsh_points():
```
The is confusing without an in argument. Or the docstring is unclear.

jorgensd replied:
There are no in-arguments as everything is stored in the `gmsh.model`.
I will try to improve the doc string.

garth-wells commented on python/demo/gmsh/demo_gmsh.py:
```diff
@@ -5,33 +5,101 @@
+
+def sort_gmsh_points():
+    """
+    Extracts geometrical nodes from all entities in gmsh
```
I don't understand this docstring.

garth-wells commented on python/dolfinx/io.py:
```diff
@@ -62,17 +62,20 @@ def read_mesh(self, ghost_mode=cpp.mesh.GhostMode.shared_facet, name="mesh", xpa
+                  5: ("hexahedron", 1), 8: ("interval", 2),
+                  9: ("triangle", 2), 10: ("quadrilateral", 2),
+                  11: ("tetrahedron", 2), 12: ("hexahedron", 2),
+                  15: ("point", 1), 21: ("triangle", 3),
```
I would have expected "point" to have degree 0.

jorgensd replied:
I can change it. For this, I was just following the previous `_gmsh_cells`.

garth-wells reviewed (changes requested):
Nice to see a change to working through the API in place of the filesystem.

francesco-ballarin commented on python/demo/gmsh/demo_gmsh.py:
```diff
@@ -5,33 +5,101 @@
+    for dim, tag in phys_grps:
+        entities = gmsh.model.getEntitiesForPhysicalGroup(dim, tag)
+        for entity in entities:
+            element_types, elem_tags, node_tags =\
```
As a followup to Garth comment. I believe that these two lines can fit in a single one (we have the limit at 120 chars, right?)

francesco-ballarin commented on python/demo/gmsh/demo_gmsh.py:
```diff
@@ -5,33 +5,101 @@
 # SPDX-License-Identifier:    LGPL-3.0-or-later
```
Are the functions in this file general enough to be included as part of the library, and not only in this demo?

jorgensd replied:
I think they eventually should be part of the library. I put them here, so that we can discuss how we would like to interact with the gmsh API.

francesco-ballarin replied:
Great. Do you and @garth-wells think that these two functions can have `gmsh.model` or `gmsh.model.mesh` as input? This would allow avoiding `import gmsh` in `dolfinx.io`.
(BTW, how would one create two meshes at the same time with such gmsh api?)

jorgensd replied:
We could put the model name as input, (then using Gmsh.model.setCurrent) such that it sets the current model, exports the geometry and topology, and returns the mesh.

jorgensd replied:
> Great. Do you and @garth-wells think that these two functions can have `gmsh.model` or `gmsh.model.mesh` as input? This would allow avoiding `import gmsh` in `dolfinx.io`.
> 
> (BTW, how would one create two meshes at the same time with such gmsh api?)

However, i am not sure how one would take a Gmsh model as input.

garth-wells replied:
Let's see how it shakes down and then decide if it can go somewhere in the library. Maybe rather than going deep into the core code we could have a `dolfinx.utilities` module or something similar.

jorgensd replied:
> Let's see how it shakes down and then decide if it can go somewhere in the library. Maybe rather than going deep into the core code we could have a `dolfinx.utilities` module or something similar.

If we do not make a type hint for the gmsh model, the only import of gmsh now is when reading in msh files.

francesco-ballarin replied:
Both `dolfinx.io` and `dolfinx.utilities` sound good to me.

jorgensd commented:
@garth-wells , @francesco-ballarin  I have now rewritten the IO such that one can either do:
1) Create the mesh in the python script and export the topology and geometry for mesh creation without any IO
2) Added IO from msh files, reading in the msh file with GMSH on processor 0, and distributing the data over the other processors. This function can also extract facet data and cell data, and create corresponding mesh tags.

garth-wells reviewed (changes requested):
Looking good, just a few more suggestions for changes.

jorgensd commented on cpp/dolfinx/io/cells.cpp:
```diff
@@ -237,11 +237,7 @@ std::vector<std::uint8_t> gmsh_tetrahedron(int num_nodes)
-    // return std::vector<std::uint8_t>{0, 1, 2, 3, 9, 6, 8, 7, 4, 5};
-    // This is the one returned by pygmsh
-    return {0, 1, 2, 3, 9, 6, 8, 7, 5, 4};
+    return std::vector<std::uint8_t>{0, 1, 2, 3, 9, 6, 8, 7, 4, 5};
```
Your right, this was just a remainder from an older iteration of this code.

francesco-ballarin commented on python/dolfinx/io.py:
```diff
@@ -62,17 +65,207 @@ def read_mesh(self, ghost_mode=cpp.mesh.GhostMode.shared_facet, name="mesh", xpa
+    return topologies
+
+
+def extract_geometry(gmsh_model, model_name=None):
```
Would not that require importing gmsh, thus making gmsh a required dependency?

jorgensd replied:
Not sure what the hint should be. To make a proper hint we would have to import gmsh at the top of the file, instead of inside `read_from_msh`.

garth-wells replied:
Indeed, normally it would require an import but to avoid that I think the type can be given as a string, e.g. `gmsh_model: "gmsh.foo".

jorgensd replied:
This doesn't pass flake8 (See failing CI).

jorgensd replied:
I will revert it if you do not have any other suggestions.

francesco-ballarin commented on docker/Dockerfile:
```diff
@@ -134,23 +133,16 @@ RUN export DEBIAN_FRONTEND=noninteractive && \
-    python3-lxml && \
-    apt-get clean && \
-    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+    libxinerama1
```
Why were the last two lines removed? I gather that they were there to reduce the docker image size.

francesco-ballarin commented on docker/Dockerfile:
```diff
@@ -134,23 +133,16 @@ RUN export DEBIAN_FRONTEND=noninteractive && \
     export HDF5_MPI="ON" && \
-    export CC=mpicc && \
-    pip3 install --no-cache-dir --no-binary=h5py h5py meshio pygmsh
+    export CC=mpicc
```
Was exporting CC only needed by h5py?

jorgensd replied:
Good question, I guess so. I'll remove them

francesco-ballarin commented on python/dolfinx/io.py:
```diff
@@ -62,17 +65,210 @@ def read_mesh(self, ghost_mode=cpp.mesh.GhostMode.shared_facet, name="mesh", xpa
+        gdim = 3
+    if MPI.COMM_WORLD.rank == 0:
+        import gmsh
+        gmsh.initialize()
```
What would happen if a user had called already `gmsh.initialize()` before calling this function?

jorgensd replied:
Fixed in latest commit

francesco-ballarin commented on python/dolfinx/io.py:
```diff
@@ -165,8 +165,14 @@ def read_from_msh(filename: str, cell_data=False, facet_data=False, gdim=None):
         import gmsh
-        gmsh.initialize()
-
+        # Check if gmsh is allready initialized
```
typo: already

francesco-ballarin commented on python/demo/gmsh/demo_gmsh.py:
```diff
@@ -4,32 +4,82 @@
+        model.addPhysicalGroup(1, [entity[1]], tag=11)
+    model.mesh.generate(2)
+    gmsh.write("mesh.msh")
+
```
Demos seems to be hanging in parallel. Is it possible that a barrier is needed here?

garth-wells replied:
Why are we writing to file?

jorgensd replied:
To show that we can read from msh.
The issue here was that I only initialized gmsh on one processor, then using it on every processor later on.

garth-wells replied:
I think it should be removed, or encapsulated in the demo. DOLFIN supports reading from XDMF. Support for other formats needs to be discussed since it's not clear where one should stop.

jorgensd replied:
I disagree, as there is no clear description of how the dolfinx-XDMF format is.
There are plenty of XDMF-files (for instance those created by meshio-convert from msh) that does not work with dolfinx. This makes it possible for any msh mesh to be loaded, and I think that should be highlighted

jorgensd replied:
I've removed them from the demo and dolfinx.io, as they would make the demo quite complex if encapsulated.

jorgensd commented:
> Looking good, just a few more suggestions for changes.

Is this ready to merge now?
